Editor: Restructure editor code

AThousandShips opened the pull request:
As discussed with @akien-mga the editor code is very cluttered, which makes it both difficult to assign code owners to, but also makes it hard to navigate

Opening as a draft to discuss the specifics of how to organize things, where I still have some areas open for discussion:
- [x] The structuring of sub folders of `/editor/`, currently these are organized mainly by type, not area, this could be changed to be more area specific (i.e. all dialogs are in `/editor/dialogs/` rather than specifically `/editor/animation/` etc.)
- [x] If some files should be moved between sub folders, for example some files that are in `/editor/gui/` that might be better in `/editor/dialogs/`
- [x] If some files that are under `/editor/plugins/` should be moved out of there and into the main sub folders, like the dialogs under `/editor/plugins/tiles/`
- [X] How and if CODEOWNERS should be restructured, currently it uses relatively rough glob patterns to handle files, this could now be made more specific (this will be fixed in a follow up PR beyond the basic changes in this, which can be done without any noise)
- [x] If we should use more relative paths, as you can see several paths were made more specific (including cases where the file references was moved to the same folder, making the specification unnecessary)

Generally looking for feedback on these specifics, but also generally, in my opinion this is badly needed to ensure the editor code remains manageable as it grows more complicated, currently the majority of files in `/editor/` lack owners, making review more complicated

---

* Supersedes: https://github.com/godotengine/godot/pull/83161
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

aaronfranke commented:
I was about to comment that we should merge #83161 before this PR, but then I saw you linking to it. Though, I still want to say that we should merge #83161 before this PR :)

AThousandShips commented:
Agreed, picked this approach at first because it was the most straightforward, and I'm leaning towards switching that out, will hold off slightly but will look at resturcturing next week probably if we agree on that switch

KoBeWi commented:
It would be nice to have some visualization of before and after the changes.
GitHub isn't very friendly for previewing the changes.

AThousandShips commented:
Can do a summary

Alex2782 commented:
What effect does it have on `scu_build`? 
If there are more subfolders, this increases the build times?

-------------


https://github.com/godotengine/godot/blob/master/scu_builders.py#L292-L303

```python
    process_folder(["editor"], ["file_system_dock", "editor_resource_preview"], 32)
    process_folder(["editor/debugger"])
    process_folder(["editor/debugger/debug_adapter"])
    process_folder(["editor/export"])
    process_folder(["editor/gui"])
    process_folder(["editor/themes"])
    process_folder(["editor/project_manager"])
    process_folder(["editor/import"])
    process_folder(["editor/import/3d"])
    process_folder(["editor/plugins"])
    process_folder(["editor/plugins/gizmos"])
    process_folder(["editor/plugins/tiles"])
```

AThousandShips commented:
Will try and compare, good point!

KoBeWi commented:
I made a before and after using this script:
```GDScript
@tool
extends Tree

@export_global_dir var base_folder: String

func _ready() -> void:
	add_folder(base_folder, null)

func add_folder(path: String, parent: TreeItem):
	var folder_item := create_item(parent)
	folder_item.set_text(0, path.get_file())
	folder_item.set_custom_color(0, Color.YELLOW)
	
	for dir in DirAccess.get_directories_at(path):
		add_folder(path.path_join(dir), folder_item)
	
	for file in DirAccess.get_files_at(path):
		if file.get_extension() != "h":
			continue
		
		var file_item := create_item(folder_item)
		file_item.set_text(0, file.get_basename())
	
	if folder_item.get_child_count() == 0:
		folder_item.free()
```
Set `base_folder` to `editor` in the source code and it will make a huge folder tree 🙃

<details>
<summary>It's really big.</summary>

| Before | After |
| --- | --- |
|![image](https://github.com/user-attachments/assets/b230e83b-05d7-4ad4-9191-643497e5ff6a)|![image](https://github.com/user-attachments/assets/5848736b-9d71-443a-92fa-48025335253e)|

</details>

I took a look at the new structure.
- There is some inconsistency between `dialogs` and `gui`. E.g. `editor_file_dialog` is in `gui`, while `create_dialog`, which is also technically a reusable GUI control, is inside `dialogs`.
  - I think `dialogs` folder might be too general. There are some dialogs that belong to docks (`connections_dialog`), to script editor (`find_in_files`), to shader editor (`editor_native_shader_source_visualiser`, I think) etc. They should be together with their base components. I'd put only "unique" and "top-level" dialogs into `dialogs` folder (`editor_about`, `editor_feature_profile` etc.)
- Should `physics` and `navigation` go under `2d`/`3d`, or should they be standalone folders? I think making them standalone makes more sense, although the current structure makes it easier to disable 2D and 3D separately.
- Texture plugins could get their own folder.
- EditorHelp stuff and DocData could be put into their own folder (together I mean).
- `project_manager` could be in `project_manager` folder.
- `project_upgrade_tool` and `3_to_4` stuff can be in a `project_upgrade` folder maybe
- `extractable_translations.gen` looks like it belongs to `translations`
- `file_info` is part of `editor_file_dialog` and `file_system_dock`. There might be better place for it than top level
- Might be worth it to add some folder in `editor_plugins` for files that aren't specific plugins? Like `editor_plugin`, `editor_plugin_settings`, but also `editor_preview_plugins` or `editor_resource_tooltip_plugins`. They look like they can be grouped together.
- `shader_globals_editor` could be in `shader` folder, except it's not part of the plugin., so not sure.
- `engine_update_label` can be inside `project manager`, it's not used anywhere else.
- `code_editor` should be in `script, although technically it's also used by shader editor. Maybe `gui` then?
- `event_listener_line_edit` should be in `gui`.
- `register_exporters` could be in `export`, unless there is a reason why it's where it is.

I could see more potential groups in the top-level folder, but it's arguable.

AThousandShips commented:
Note that the dialogs folder has been reverted

AThousandShips commented:
There, made a major restructure, dropped the old "dialogs" division, and moved into a few new categories

Still to do:
* [ ] Decide on moving `editor/plugins/editor_plugin.h` and similar, this is used *everywhere* and is also used in custom modules, so it's risky, and the returns are limited, but it would improve organization
* [ ] Decide on any missing categories, and some division between them
* [ ] Move some code from `editor/SCsub` to other parts, the script parts generating documentation etc.
* [ ] Figure out some further specifics on `CODEOWNERS`

Otherwise this is nearly in what I'd prefer as the final state

CC @KoBeWi @akien-mga @Repiteo

KoBeWi commented:
The new structure:
<details>
<summary>Unfold</summary>

![image](https://github.com/user-attachments/assets/d2ea53b5-9e18-4165-8b99-47bb17abdfdb)

</details>

- Not sure about `textures/`, it sounds like a folder with textures. `texture/` could be better, if it's fine.
- `color_channel_selector` should be in `textures/`.
- `canvas_item_editor_plugin` should be in `plugins/2d`.
- Some `.gen` files weren't moved to proper subfolders and are still in `editor/`
- `editor_build_profile` could go to `settings/`

Also some of my previous comments still apply, but overall this new structure looks much better.

AThousandShips commented:
> Some .gen files weren't moved to proper subfolders and are still in `editor/`

Have you made sure to clean before building? They have been moved

> `canvas_item_editor_plugin` should be in `plugins/2d`.

It belongs in the base as it isn't 2D *or* GUI, same reasoning why `canvas_item` was moved to `scene/main` for 4.0

AThousandShips commented:
For the prioritizing physics vs 2D/3D I prefer keeping it as is, or possibly adding a sub-folder, I prefer having things closer to the way `scene` is organized, it helps navigation

KoBeWi commented:
> Have you made sure to clean before building? They have been moved

I didn't. I actually didn't even build, so this is probably why 🙃

> Can you please point to which, I've resolved most or responded reasons why they weren't

I crossed out resolved stuff.

AThousandShips commented:
Will update! `doc_data_compressed.gen` must have been a mistake don't remember moving it there!

Audio makes sense, two files should be enough to move, in both plugins and the root, will update and rebase soon

AThousandShips commented:
Moved the suggested, as well as audio both in `editor/` and `editor/plugins/`, as well as `credits_roll` into `editor/gui/`

Added scu for `editor/plugins/script/` which was excluded previously, for these changes my condition is three or more files to justify it, will compare build times for SCU builds later today probably

KoBeWi reviewed (approved):
Final structure:
<details>
<summary>Unfold</summary>

![image](https://github.com/user-attachments/assets/4db3e5f5-f03d-460a-9efb-c4ad0108a00f)

</details>

It's good enough. I wonder about `plugins/themes/` folder (could be merged with `plugins/gui/`), but it's probably fine.

AThousandShips commented:
That's a good point, will merge those

AThousandShips committed:
Editor: Restructure editor code

Moving various editor files into sub folders to reduce clutter

Alex2782 commented:
There is this description, which will probably no longer be current after PR merge.
https://docs.godotengine.org/en/stable/contributing/development/editor/introduction_to_editor_development.html

_(Last updated 2 years ago)_

AThousandShips commented:
I incorporated the new changes, plugins is now essentially clean, and the structure is far more organized, and things are more or less in one place, i.e. there's no separation between plugins and other folders for the same area like animation

Some things could still be restructured but this should be a pretty good start, some content in `editor/gui/` could still be divided out, but I think this is a good point

AThousandShips commented on .github/CODEOWNERS:
```diff
@@ -44,18 +44,22 @@
+/editor/audio/                                @godotengine/audio
 /editor/debugger/                             @godotengine/debugger
+/editor/doc/                                  @godotengine/documentation
+/editor/docks                                 @godotengine/docks
```
```suggestion
/editor/docks/                                @godotengine/docks
```

KoBeWi commented:
Latest structure for reference:

<details>
<summary>Unfold</summary>

![image](https://github.com/user-attachments/assets/c6ddb420-7202-47ae-a5bf-41df4e77d58a)

</details>

Repiteo reviewed (approved):
This folder has been in DIRE need of this sort of restructuring, and I'm thrilled with this implementation! My only possible reservation would be moving `embeeded_process` from the editor root to the `run` subfolder, but otherwise I don't want to bikeshed this any longer than it already has. Great work!

AThousandShips commented:
Will rebase and fix the final details! Will be great to have this done, but it's been an interesting experience and I now know the editor even better!

Repiteo commented:
Thanks!

AThousandShips commented:
Thank *you*! Now we reap the rewards!
